tutorials/haskell/snap-web-app/tutorial.md

```diff
@@ -0,0 +1,479 @@
+it has a unique view which consists in a Todo items list with a couple of buttons
+to mark it as done or delete it, and a form for new items submission purpose. Just
+like this
+
+<p align="center">
+  <img alt="Snap Todo web app" src="todo.png" style="width:50%;height:50%;" />
+  </br>
+</p>
+
+
+This application uses Snap and therefore it seemed logical to use its default template engine: Heist.
+
+Heist uses simple XML tags to bind values to your templates in a straightforward
+way. It's a great library which even allows to use compiled or interpreted templates,
+but that's out of the scope of the tutorial. Anyway, you should know this
+tutorial uses interpreted templates, and also another structure called [Splice](http://snapframework.com/docs/tutorials/heist#heist-programming) to pull
+values from. You'll see more about that in a bit
+
+This is an extract of how the [todos.tpl](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/snaplets/heist/templates/todos.tpl) Heist template for this tutorial
+application looks like. The submission form was left out for the sake of simplicity
+
+```html
+<apply template="base">
+  <div class="col-xs-6 col-xs-offset-3">
+    <div class="well">
+      <h1>TODO Items</h1>
+      <ul class="list-group">
+        <todos>
+          <input type="hidden" value="${id}">
+          <li class="list-group-item clearfix ${successClass}">
+            <span><description/></span>
+            <div class="pull-right" role="group">
+              <a class="btn btn-xs btn-success img-circle"
+                  onclick="document.getElementById('form-check-${id}').submit();">
+                  ✓
+              </a>
+              <form id="form-check-${id}" action="/todos/${id}/complete/${notCompleted}" method="post" class="hidden-form">
+              </form>
+              <span> </span>
+              <a class="btn btn-xs btn-danger img-circle"
+                  onclick="document.getElementById('form-rmv-${id}').submit();">
+                Ｘ
+              </a>
+              <form id="form-rmv-${id}" action="/todos/${id}/delete" method="post" class="hidden-form">
+              </form>
+            </div>
+          </li>
+        </todos>
+      </ul>
+      <div>
+        <hr>
+        <div class="clearfix">
+          <!-- forms goes here  -->
+        </div>
+      </div>
+    </div>
+  </div>
+</apply>
+```
+Lets explain this a little bit. The `<apply template="base">` tag means this
+view uses a layout Heist template called `base.tpl`, you can have a closer look
+to it [here](https://github.com/stackbuilders/tutorials/tutorials/haskell/snap-web-app/code/snaplets/heist/templates/todos.tpl). The `<todos>` node implies everything it encloses will be repeated
+for every single Todo item. That comes from the Todo index request handler
+
+```haskell
+handleIndexTodo :: Handler App App ()
+handleIndexTodo = do
+  c <- gets _conn
```

Comment:
This isn't wrong, but since you already have lenses you can `over` instead of `gets`.